### ISSUE TITLE
SalePrice / ProductPrice

### ISSUE THREAD

**Issue by crls12opazo** [opensourcepos/opensourcepos#1647]
New Feature / Enhancement
 
                        Dear community,

I propose that when updating the price in the sales chart, this makes the calculation for the quantity. In order to improve the sales module and also that the community can do it.
As is done now with quantity which changes the price according to the result. This can also be implemented in the event that the price of the sale changes.

This would help a lot for those who make small sales and have to type the weight of the products according to the balance. It is easier to enter the price of the sale instead of the grams.
The modification must be only in the sale and this should not modify the selling price of the article.

The formula would be
(SalePrice / ProductPrice) this would result the quantity as a fraction.

Greetings.

**Comment by SteveIreland** [comment]
I'm a bit confused by the request and it's probably due to terminology/language mapping challenges.  So to clarify, do you want to be able to enter the "Total" and have the quantity derived from Total divided by Price?

It sounds simple enough, but I don't think I would want the total amount to be input capable for most of my items.  So maybe this feature is driven by item type or a system configuration option.  I would lean toward an item type of "Weight-based".

**Comment by crls12opazo** [comment]
your suggestion is better

**Comment by daN4cat** [comment]
That requirement doesn't really make sense to me. How do you monitor your inventory if you go by extrapolation? You sell by price instead by real number of pieces?
Unless you mean like a candy shop where you buy $10 and instead of counting you check how many using OSPOS, but I would say use a scale and work by weight which is probably better.

**Comment by nathanzky** [comment]
crls12opazo suggestion is the same as my requirement I would like to test the OSPOS running a petrol gas station let say the customer need to load USD 50. Instead of entering how many liters you enter the amount and it automatically calculates how many liters it is.. Would this be possible?

**Comment by mckaygerhard** [comment]
same for the stores that sale fruits by exampel a client need "20 $ for stranberry" event a "1.2 kg of stranberry"

**Comment by SteveIreland** [comment]
I'd like to take a stab at this to see if a practical solution can be put together for this.

**Comment by crls12opazo** [comment]
thank you!!

**Comment by SteveIreland** [comment]
It's in progress.  

I believe this should be an optional feature called "Price Based Sales".   

If enabled then a new item type of "Price Based" will be available for selection.  

A "Price Based" item will be treated like a "Standard" item except that the option of entering an extended total will be available.  The extended line item total on the register will be input capable (which is currently output only).  

If the extended amount is entered then the system will recalculate the quantity of the item sold -
 rounding it to the number of decimals that the system is configured for.

**Comment by SteveIreland** [comment]
If I'm off track someone stop me

**Comment by mckaygerhard** [comment]
well @SteveIreland your comments are currently the only options here.. so why stop!?
about the suggestion feature, seems a lot of work in js/jquery (that i hate it je je) but make sense in that way

**Comment by daN4cat** [comment]
@SteveIreland sounds sensible to me and yes make an optional story or maybe should be another type of item, so more an option at item level.

**Comment by SteveIreland** [comment]
I've submitted a pull request for this.  If you're interested in the feature please give it a spin and let me know if there are any questions about it or problems with it.  I'll be happy to fix whatever's wrong.

To use this feature there is a new option named "Allow Derived Sale Quantity"  You'll need to check that option.

Then for those items that would be helpful to be added to a sales by entering an extended sales amount instead of a quantity you will need to set the item type to "Amount Entry".

Then for a sale when that item type is selected the extended line amount will become input capable and the extended line amount is changed the the quantity should be recalculated.

This does NOT prevent you from entering the quantity directly.

Note that the extended amount that you enter might change in the situation that the price multiplied by the calculated quantity (due to rounding issues) is different than the amount you enter.

Enjoy!

**Comment by daN4cat** [comment]
Now in master, @crls12opazo please check and test the latest code and see if the feature is as per your expectations.

Thanks @SteveIreland for the help and contribution. I made this feature part of 3.2.0 as it looks to me not too much of an impact... pending some test though.

**Comment by odiea** [comment]
Needs more work. Should also include the app config enabled and it's missing the total with discount applied.
```
if($item['item_type'] == 2)
									{
									echo form_input(array('name'=>'total', 'class'=>'form-control input-sm', 'value'=>to_currency_no_money($item['total']), 'tabindex'=>++$tabindex));
									}

**Comment by daN4cat** [comment]
OK, there is more to fix than what I did. @SteveIreland please check as I need to head out now.

**Comment by SteveIreland** [comment]
I'm not yet clear on what the issue is, so I don't want to dive back in there until I understand the issue.

Regarding @odiea 's  concern about needing to add the app config option I suppose that's a matter of programming style and preference.  I feel it's not necessary because if the item type is 2 then that is what should drive the logic.  It would, in my opinion, be redundant.  One cannot set the item type to 2 without the feature being enabled.  
But it IS possible to have an item type set to 2 and have the feature disabled.  But that would be bad, bad, bad.  If you enable the feature, set the flag to 2 and then disable the feature then you would have essentially stepped on your own toe.  The proper recovery would be to re-enable the feature and set the item types back to "Standard" and then disable the feature.  That's the only way to not have a bad data scenario.  

If I allowed the item type 2 continue to be 2 and NOT implement the expected behavior for item type 2 then you'll have a potential problem with other parts of the code that expect the standard item type of 1 and no one would be easily able to explain the problem because you will no longer SEE that the item type has been set to "2".

At this point I still feel pretty strongly that the code should remain as it was originally coded.

However, if you tell me you still want it changed then let me know.

**Comment by odiea** [comment]
This is your baby. My major concern was the missing discount being applied and shown on the register. I then noticed that the 2 item feature was still showing after the feature was disabled.

**Comment by odiea** [comment]
After using it I can also see that this may be a benefit if just the manager has the ability to change this feature and enable it for some items then disable it so others cannot add this feature to other Items.

**Comment by daN4cat** [comment]
So it sounds like my last fix was the real problem and all the rest works fine. Let's wait for a bit of more testing before we can close.

**Comment by odiea** [comment]
Yes. I would say all is fine now as far as what I was concerned about. Happy New Year.

**Comment by SteveIreland** [comment]
Sooo .... before we leave this let's talk about the missing discount because I'm missing where the discount is missing and that concerns me too.

It is my understanding that the discount has NOT ever been applied to the extended line amount.  It IS tracked internally but not reflected in the extended line item total.   As a result (of that assumption) I am not currently factoring in any discount.  The discount is recalculated after I recalculate the quantity and the line total (excluding discount) and the discounted is reflected in the sales subtotals.

I tested the applied discount and it seems to be working (as I expected) but if it's not working as you expected please let me know and help me understand with an example.

**Comment by odiea** [comment]
This missing line of code( goes above the javascript line) updated the line item total after the discount was applied.
```
<td><?php echo to_currency($item['price']*$item['quantity']-$item['price']*$item['quantity']*$item['discount']/100); ?></td>

<td><?php echo form_input(array('name'=>'discount', 'class'=>'form-control input-sm', 'value'=>to_decimals($item['discount'], 0), 'tabindex'=>++$tabindex));?></td>
							<td>
								<?php
								if($item['item_type'] == 2) {
									echo form_input(array('name'=>'total', 'class'=>'form-control input-sm', 'value'=>to_currency_no_money($item['total']), 'tabindex'=>++$tabindex));
								} else {
									echo to_currency($item['total']);
								}
								?>
							</td>
							<td><a href="javascript:document.getElementById('<?php echo 'cart_'.$line ?>').submit();" title=<?php echo $this->lang->line('sales_update')?> ><span class="glyphicon glyphicon-refresh"></span></a></td>
							</tr>
```

**Comment by daN4cat** [comment]
Yes, I copy and pasted the old line into the "else" part to avoid any regression. Not sure how that works with the new calculation based on weight. I guess the discount needs to be applied so there is bug with that regards.

**Comment by odiea** [comment]
The discount is applied to the new calculation. It shows on the right hand side as a discounted price and on the receipt ok but not on the Register where we are familiar with i showing.

**Comment by SteveIreland** [comment]
Okay, the "oh duh" moment finally kicked in and I acknowledge the error of my way.  Thanks for your patience.

I'm waffling here since I don't know the "real" business rules, but I'm assuming the amount to be entered will NOT have the discount pre-applied.  So here is the rule I'm going to move forward with and if anyone disagrees please advise.

If the item type is "Amount Entry" then the amount to be entered for the sale amount will be pre-discount.  Oddly enough I will compare that with against the post-discount total and if there is a difference I will recalculate the quantity using the amount entered.  The amount entered will also be changed automatically to the "post discount total" and that is the amount that will be displayed.

**Comment by odiea** [comment]
You can see here that the discount is being applied to the Sale. It just does not appear in the Item table which is not an issue since it is now a input.
![2018-01-01 12_31_30-wedidit _ powered by ospos 3 2 0](https://user-images.githubusercontent.com/10962177/34470518-16f18dec-eef0-11e7-9717-65ccbf4197be.png)

**Comment by SteveIreland** [comment]
So you think it is okay as is without the discount being applied?

If the sale contains both types of items and each has a discount then one Total will include the discount and the other won't.  I think someone might eventually complain about that but I'm okay with waiting.

I think I still need to submit a followup because somewhere in the merge the creation of the new configuration value was dropped from the tables.sql, and the database.sql.  

It looks like that might have been intentional so I"ll wait for @daN4cat to advise.

**Comment by daN4cat** [comment]
I would calculate the price based on the quantity pre-discount and then apply the discount to get the line total.

I dropped the .sql because now the migrate does everything. We don't need to maintain those tables anymore as the automigrate does everything post 3.0.0.

**Comment by SteveIreland** [comment]
For a new install I manually run the database.sql.  If the new config value isn't in the database.sql script where/when will it be added?

**Comment by daN4cat** [comment]
I updated application/migrations/sqlscripts/3.1.1_to_3.2.0.sql and renamed application/migrations/<date>_upgrade_to_3.2.0.php to have today's date so it's run again.

**Comment by SteveIreland** [comment]
Okay.  I've made a few tweeks and submitted a new pull request for this change.

Now, the "discounted extended line amount" should be being displayed for each sale line.

When the user wants to enter the line amount they only need to enter the non-discounted extended line amount in the total column.  It will recalculate the quantity based on the amount entered and then recalculate the extended line total (applying the discount).

@daN4cat I'm still unclear on a point.  When creating a database from scratch, do we now run the existing database.sql and then expect the migration script to add new database changes (including priming the pump for new configuration values)?

I can see how this would work but that assumes the current database.php will be the starting point for all future database changes.  Is that the intent?

**Comment by SteveIreland** [comment]
Okay, I'm trying to anticipate your intent.  :-)  When you publish a new release will you be taking the "upgrade to" scripts and manually merging them into the database.php?

**Comment by daN4cat** [comment]
You use the current database.sql and the upgrade script runs the first time you log in OSPOS upgrading the database using the migration scripts.

**Comment by daN4cat** [comment]
WRT your latest change, I'm afraid I don't get where the discounted line is calculated now...

**Comment by SteveIreland** [comment]
Sale_lib.php line 903.  This was existing code that I thought I'd take advantage of.

**Comment by odiea** [comment]
I copied and pasted the 3 changed file this am. Here are the results. The 2 item edits ok with the discount but a normal item does not and gives the following error.
![2018-01-02 06_54_24-wedidit _ powered by ospos 3 2 0](https://user-images.githubusercontent.com/10962177/34485878-07c55b38-ef8a-11e7-9094-b35753af8947.png)

**Comment by odiea** [comment]
I cannot edit price, quantity or discount now on a standard item

**Comment by SteveIreland** [comment]
Not good.  Not good at all.   I'll check to see why.  I must have mucked up a last second change I made to try to flag a discounted total as required for item type 2.

**Comment by SteveIreland** [comment]
Yep ... that's what happened.  The quick resolution is to remove the test.  It was working before I added it and if they enter in a zero extended amount it simply results in a zero quantity being calculated.

I don't know if it requires validation.   Thoughts?

**Comment by SteveIreland** [comment]
I could go retrieve the item type, but that would require another database access for something that isn't going be widely used.

**Comment by odiea** [comment]
That's a question for @daN4cat ,@jekkos. I just test.

**Comment by SteveIreland** [comment]
And I just code.   We need a decider. :-)

**Comment by odiea** [comment]
Here is my simple code that appears to work with your original commit.
```
<td><?php echo form_input(array('name'=>'discount', 'class'=>'form-control input-sm', 'value'=>to_decimals($item['discount'], 0), 'tabindex'=>++$tabindex));?></td>
							<td>
								<?php
								if($item['item_type'] == 2 && $this->config->item('derive_sale_quantity') == '1')
									{
									$item['total'] = ($item['price']*$item['quantity']-$item['price']*$item['quantity']*$item['discount']/100);
									echo form_input(array('name'=>'total', 'class'=>'form-control input-sm', 'value'=>to_currency_no_money($item['total']), 'tabindex'=>++$tabindex));
									} 
								else
									{
								 echo to_currency($item['price']*$item['quantity']-$item['price']*$item['quantity']*$item['discount']/100); 
									}
								?>
							</td>

**Comment by SteveIreland** [comment]
That's not the current issue.  It's some code in Sales.php that is causing the current problem.

**Comment by SteveIreland** [comment]
I'll go ahead and remove the problematical validation within the next hour if I don't hear from a decider or original reqestor.

**Comment by SteveIreland** [comment]
Done .. the pull request has been updated.

**Comment by SteveIreland** [comment]
Thanks @odiea

**Comment by daN4cat** [comment]
In master now.

**Comment by SteveIreland** [comment]
Thanks @daN4cat and good catch on the hard-coded item type.

**Comment by daN4cat** [comment]
@crls12opazo @nathanzky any feedback? Did you find the time to test it?

**Comment by SteveIreland** [comment]
@crls12opazo  What were the problems?

**Comment by crls12opazo** [comment]
![captura](https://user-images.githubusercontent.com/32281096/34561843-a1aced82-f12a-11e7-8682-43eea012ae9b.JPG)

**Comment by odiea** [comment]
You downloaded from the Github master and not from the Local bintray install . This may be the most current version. https://bintray.com/jekkos/opensourcepos/download_file?file_path=opensourcepos.20180102222709.master.8b96af.zip

**Comment by crls12opazo** [comment]
![image](https://user-images.githubusercontent.com/32281096/34562560-bd636cd8-f12d-11e7-9064-d7d87f9d3fb4.png)
![image](https://user-images.githubusercontent.com/32281096/34562579-d3cd6e4c-f12d-11e7-85c6-3b170be37d93.png)
here the quantity must be 0.5

**Comment by odiea** [comment]
You will need to change the total amount to 50.00 to get the 0.5. You also need to change the quantity decimals.

**Comment by crls12opazo** [comment]
![image](https://user-images.githubusercontent.com/32281096/34563207-6f631562-f130-11e7-9e5a-54b5d874f711.png)
perfect! , the price textbox must be disabled, thank you @odiea

**Comment by odiea** [comment]
A quick primer on the new function.
Go to Store config. Enable the following.
![2018-01-04 05_22_41-wedidit _ powered by ospos 3 2 0](https://user-images.githubusercontent.com/10962177/34563460-fb44bc3c-f10f-11e7-93a6-53f97cace311.png)
Now set the quantity decimals as needed.
![2018-01-04 05_24_15-wedidit _ powered by ospos 3 2 0](https://user-images.githubusercontent.com/10962177/34563492-164666e8-f110-11e7-91ee-e2f214643809.png)
In Items a new radio button will show and choose the new button.
![2018-01-04 05_23_26-wedidit _ powered by ospos 3 2 0](https://user-images.githubusercontent.com/10962177/34563439-e1b31318-f10f-11e7-8976-ea68ca2980b2.png)
Next in Sales change the  total amount as needed.
![2018-01-04 05_24_49-wedidit _ powered by ospos 3 2 0](https://user-images.githubusercontent.com/10962177/34563599-6b9f7cce-f110-11e7-99e4-88fc776ade22.png)

**Comment by crls12opazo** [comment]
i do

**Comment by daN4cat** [comment]
So is it fine?

**Comment by crls12opazo** [comment]
all fine, thank to all!

**Comment by nathanzky** [comment]
Thanks to all
